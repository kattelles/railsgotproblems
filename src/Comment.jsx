import React, { Component } from 'react'
import Markdown from 'react-markdown'
import { AddLinkstoAtSymbols } from './utils/utils'

class Comment extends Component {
  constructor() {
    super();
    this.state = {
      userUrl: '',
      userName: '',
      userAvatar: '',
      body: ''
    };
  }

  componentDidMount() {
    let comment = this.props;
    let body = AddLinkstoAtSymbols(comment.body);
    this.setState({
      userUrl: comment.userUrl,
      userName: comment.userName,
      userAvatar: comment.userAvatar,
      body: body
    });
  }

  render() {
    return (
      <div className='comment'>
        <div className='avatar'>
          <a href={this.state.userUrl} target='_blank'>
            <img alt='avatar' src={this.state.userAvatar}/>
          </a>
          <a href={this.state.userUrl} target='_blank'>
            @{this.state.userName}
          </a>
        </div>
        <Markdown className='markdown comment-body' source={this.state.body}/>
      </div>
    );
  }
}

export default Comment;
