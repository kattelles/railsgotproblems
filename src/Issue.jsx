import React, { Component } from 'react'
import { browserHistory as history} from 'react-router'
import Markdown from 'react-markdown'
import { AddLinkstoAtSymbols, FindAllValuesInString } from './utils/utils'

class Issue extends Component {
  constructor() {
    super();
    this.state = {
      issueNumber: '',
      issueTitle: '',
      labels: [],
      reporterName: '',
      reporterGravatar: '',
      summary: ''
    };

    this.viewDetails = this.viewDetails.bind(this);
  }

  componentDidMount() {
    let issue = this.props.issue;
    let summary = this.parseSummary(issue.body);
    let title = this.parseTitle(issue.title);
    this.setState({
      issueNumber: issue.number,
      url: issue.html_url,
      issueTitle: title,
      labels: issue.labels,
      reporterName: issue.user.login,
      reporterGravatar: issue.user.avatar_url,
      reporterUrl: issue.user.html_url,
      summary: summary
    });
  }

  parseTitle(title) {
    // hack to fix prevent long titles without spaces from overflowing
    if (title.substr(0, 50).indexOf(' ') === -1) {
      return title.substr(0, 50) + '...';
    }
    return title;
  }

  parseSummary(summary) {
    let shortSummary = summary.slice(0, 140);
    let parsedSummary = '';
    let periods = FindAllValuesInString(shortSummary, '.');
    if (periods.length > 0) {
      let periodIdx = periods.pop();
      parsedSummary = shortSummary.slice(0, periodIdx) + '...';
    } else {
      let spaces= FindAllValuesInString(shortSummary, ' ');
      let spaceIdx = spaces.pop();
      parsedSummary = shortSummary.slice(0, spaceIdx) + '...';
    }
    if (parsedSummary.length > 0) {
      return AddLinkstoAtSymbols(parsedSummary);
    }
    return AddLinkstoAtSymbols(shortSummary);
  }

  viewDetails(event) {
    history.push(`/issue/${this.state.issueNumber}`);
  }

  render() {
    let labels = this.state.labels.map(label => {
      let labelUrl = `https://github.com/search?q=topic%3A${label.name}+org%3Aruby&type=Issues`;
      return(<div className='issue-label'
              key={label.id}>
              <a href={labelUrl}
                  className='issue-label'
                  target='_blank'>
                {label.name}
              </a>
            </div>);
    });

    return (
      <div onClick={this.viewDetails} className='issue'>
        <div className='avatar'>
          <a href={this.state.url}
            target='_blank'
            className='issue-number'>
            #{this.state.issueNumber}
          </a>
          <a href={this.state.reporterUrl} target='_blank'>
            <img alt='avatar' src={this.state.reporterGravatar}/>
          </a>
          <a href={this.state.reporterUrl} target='_blank'>
            @{this.state.reporterName}
          </a>
        </div>
        <div className='issue-right'>
          <div className='issue-title'>{this.state.issueTitle}</div>
          <div className='labels-group'>{labels}</div>
          <Markdown
            className='issue-summary markdown'
            source={this.state.summary}/>
        </div>
      </div>
    );
  }
}

export default Issue;
