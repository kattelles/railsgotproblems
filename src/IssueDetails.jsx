import React, { Component } from 'react'
import { browserHistory as history} from 'react-router'
import Comment from './Comment'
import Markdown from 'react-markdown'
import Loader from 'react-loader'
import { AddLinkstoAtSymbols } from './utils/utils'

class IssueDetails extends Component {
  constructor() {
    super();
    this.state = {
      title: '',
      state: '',
      reporterName: '',
      reporterGravatar: '',
      summary: '',
      comments: [],
      loaded: false,
    };
  }

  componentDidMount() {
    let issueNum = this.props.params.number;
    let url = `https://api.github.com/repos/rails/rails/issues/${issueNum}`;
    let commentUrl = `https://api.github.com/repos/rails/rails/issues/${issueNum}/comments`;
    this.fetchIssue(url);
    this.fetchComments(commentUrl);
  }

  fetchIssue(url) {
    fetch(url)
      .then(response => response.json())
      .then(issue =>  {
        let summary = AddLinkstoAtSymbols(issue.body);
        this.setState({
          title: issue.title,
          state: issue.state.toUpperCase(),
          createdAt: issue.created_at.substr(0, 10),
          updatedAt: issue.updated_at.substr(0, 10),
          reporterName: issue.user.login,
          reporterGravatar: issue.user.avatar_url,
          reporterUrl: issue.user.html_url,
          summary: summary,
          loaded: true
        });
      });
  }

  fetchComments(url) {
    fetch(url)
      .then(response => response.json())
      .then(comments =>  {
        this.setState({comments: comments});
      });
  }

  backButton() {
    history.push(`/`);
  }

  render() {
    let comments = this.state.comments.map(comment => {
      return(
        <Comment
          key={comment.id}
          userUrl={comment.user.html_url}
          userName={comment.user.login}
          userAvatar={comment.user.avatar_url}
          body={comment.body}/>
      );
    });

    let commentsHeader = '';
    if (comments.length > 0) {
      commentsHeader = (<div className='comments-header'>Comments</div>);
    }

    return (
      <div className='issue-details-outer'>
        <Loader color='#ff3b3f' loaded={this.state.loaded}>
          <div className='issue-details'>
            <div className='avatar'>
              <a href={this.state.reporterUrl} target='_blank'>
                <img src={this.state.reporterGravatar}
                      width='100'
                      height='100'
                      alt='avatar'/>
              </a>
              <a href={this.state.reporterUrl} target='_blank'>
                @{this.state.reporterName}
              </a>
            </div>
            <div className='issue-details-header'>
              <div className='issue-title'>{this.state.title}</div>
              <div className='issue-details-date'>
                Created: {this.state.createdAt}
              </div>
              <div className='issue-details-date'>
                Last updated: {this.state.updatedAt}
              </div>
              <div className='issue-details-state'>{this.state.state}</div>
            </div>
          </div>
          <Markdown className='markdown' source={this.state.summary} />
          {commentsHeader}
          {comments}
          <div className='back-button button' onClick={this.backButton}>
            back
          </div>
        </Loader>
      </div>
    );
  }
}

export default IssueDetails;
