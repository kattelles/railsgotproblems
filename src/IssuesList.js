import React, { Component } from 'react'
import Parse from 'parse-link-header'
import Issue from './Issue'
import Loader from 'react-loader'

class IssuesList extends Component {
  constructor() {
    super();
    this.state = {
      issues: [],
      headers: {},
      lastPageReached: false,
      loaded: false
    };

    this.getMoreResults = this.getMoreResults.bind(this);
    this.makeApiCall = this.makeApiCall.bind(this);
  }

  componentDidMount() {
    // initial api call
    this.makeApiCall(`https://api.github.com/repos/rails/rails/issues?&per_page=25&page=1`);
  }

  makeApiCall(url) {
    fetch(url)
      .then(response => {
        let headers = response.headers.get('Link');
        let parsedHeaders = Parse(headers);
        this.setState({headers: parsedHeaders});
        return response.json();
      }).then(results =>  {
        this.setState({
          issues: this.state.issues.concat(results),
          loaded: true
        });
      });
  }

  getMoreResults() {
    let lastPage = this.state.headers.last;
    let nextPage = this.state.headers.next;
    if (nextPage.page === lastPage.page ) {
      this.setState({lastPageReached: true});
    }
    this.makeApiCall(nextPage.url);
  };

  render() {
    let issues = this.state.issues.map(issue => {
      return (<Issue key={issue.id} issue={issue}/>);
    });

    let getMoreResults = '';
    if (this.state.lastPageReached) {
      getMoreResults =
        (<img
          alt='party-parrot'
          className='party-parrot'
          src={require('./../public/img/parrot.gif')}/>);
    } else {
      getMoreResults = (
        <div
          className='show-more-results button'
          onClick={this.getMoreResults}>
          Show more results
        </div>);
    }

    return (
      <Loader className='app' color='#ff3b3f' loaded={this.state.loaded}>
        {issues}
        {getMoreResults}
      </Loader>
    );
  }
}

export default IssuesList;
