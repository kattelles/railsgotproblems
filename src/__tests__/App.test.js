import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

// components
import IssuesList from '../IssuesList';
import Issue from '../Issue';
import IssueDetails from '../IssueDetails';
import Comment from '../Comment';

// smoke test
it('app renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<IssuesList />, div);
});

// snapshot tests
describe(IssuesList, () => {
  it('renders IssuesList and matches snapshot', () => {
    const component = renderer.create(
      <IssuesList />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe(Issue, () => {
  it('renders Issue and matches snapshot', () => {
    const component = renderer.create(
      <Issue/>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe(IssueDetails, () => {
  it('renders IssueDetails and matches snapshot', () => {
    let params = {number: '28160'};
    const component = renderer.create(
      <IssueDetails params={params}/>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe(Comment, () => {
  it('renders Comment and matches snapshot', () => {
    const component = renderer.create(
      <Comment />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
