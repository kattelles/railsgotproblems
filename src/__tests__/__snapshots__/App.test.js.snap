exports[`class Comment extends /* istanbul ignore next */_react.Component {
  constructor() {
    super();
    this.state = {
      userUrl: "",
      userName: "",
      userAvatar: "",
      body: "" };

  }

  componentDidMount() {
    let comment = this.props;
    this.setState({
      userUrl: comment.userUrl,
      userName: comment.userName,
      userAvatar: comment.userAvatar,
      body: comment.body });

  }

  render() {
    return (
      /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'div', /* istanbul ignore next */{ className: 'comment', __source: { fileName: _jsxFileName, lineNumber: 27 }, __self: this },
        /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'div', /* istanbul ignore next */{ className: 'avatar', __source: { fileName: _jsxFileName, lineNumber: 28 }, __self: this },
          /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'img', /* istanbul ignore next */{ role: 'presentation', src: this.state.userAvatar, __source: { fileName: _jsxFileName, lineNumber: 29 }, __self: this }),
          /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'a', /* istanbul ignore next */{ href: this.state.userUrl, target: '_blank', __source: { fileName: _jsxFileName, lineNumber: 30 }, __self: this }, /* istanbul ignore next */'@', this.state.userName)),

        /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */_reactMarkdown2.default, /* istanbul ignore next */{ className: 'markdown comment-body', source: this.state.body, __source: { fileName: _jsxFileName, lineNumber: 32 }, __self: this })));


  }} renders Comment and matches snapshot 1`] = `
<div
  className="comment">
  <div
    className="avatar">
    <img
      role="presentation"
      src={undefined} />
    <a
      href={undefined}
      target="_blank">
      @
    </a>
  </div>
  <div
    className="markdown comment-body" />
</div>
`;
